recipe/2024-06-private-bicep-registries.md

```diff
@@ -0,0 +1,425 @@
+```
+*Admin User Credentials*:
+
+Users can enable the admin user and obtain these credentials from the Azure portal:
+```
+Username : The ACR admin username.
+Password : The ACR admin password.
+```
+
+*Azure CLI Token*:
+```
+Username: Use '00000000-0000-0000-0000-000000000000' (a placeholder indicating token-based authentication).
+Password: Use the token obtained from az acr login.
+```
+
+You can obtain the token using the Azure CLI:
+```
+az acr login --name <registry-name> --expose-token
+```
+**Github**:
+
+Github uses personal access token as password to auth into ghcr.
+```
+Username: Github username.
+Password: Personal access token.
+```
+
+**AWS**:
+```
+Username: AWS
+Password: output of `aws ecr get-login-password --region region`
+```
+While there are various container registry providers, each offering multiple authentication methods, by leveraging ORAS, we abstract away the need for provider-specific code to handle authentication. 
+
+ORAS auth client example to authenticate private registry using username and password,
+```
+	repo.Client = &auth.Client{
+		Client: retry.DefaultClient,
+		Credential: auth.StaticCredential("orasregistry.azurecr.io", auth.Credential{
+			Username: "<username>",
+			Password: "<password>",
+		}),
+	}
+	repo.Client = client
+```
+
+#### Authentication using Federated identity
+**Azure**:
+
+Users can leverage Azure Federated Identity to authenticate seamlessly. When Azure Federated Identity is set up on a Kubernetes cluster, it enables secure and managed authentication, allowing workloads running on the cluster to access ACR without needing to manage service principal credentials directly. This approach enhances security by eliminating the need for hardcoded credentials and simplifies the authentication process by relying on Azure Active Directory (AAD) token issuance for the cluster's managed identity.
+
+[Enable Azure Workload Identity on Radius:](https://docs.radapp.io/guides/operations/providers/azure-provider/howto-azure-provider-wi/)
+```
+rad install kubernetes --set global.azureWorkloadIdentity.enabled=true
+```
+or by configuring azure provider with Azure Credential kind `Workload Identity` using `rad init --full` command.
+
+To authenticate ACR using Azure Federated credentials, we need to get the refresh token from ACR, and use it with ORAS auth client. The detailed steps are outlined below:
+
+- Retrieve the Azure Active Directory (AAD) token for the client ID configured with Azure Federated Identity.
+- Get refresh token from ACR by exchanging for the above AAD access token, set the 
+  ```
+  formData := url.Values{
+      "grant_type":   {"refresh_token"},
+      "service":      test.azurecr.io,
+      "tenant":       <tenant-id>,
+      "access_token": <aad-token>,
+      "expires_in": <5 mins in seconds> 
+    }
+    
+    // jsonResponse contains the refresh token from ACR.
+    jsonResponse, err := http.PostForm("https://test.azurecr.io/oauth2/exchange", formData)
+  ```
```

Comment:
A few questions about this ... does MSAL provide a library for this? I'm a little bit anxious about doing the raw HTTP calls ourselves if we don't have to since it's security sensitive.

Is this based on a sample the ACR team provided? or just something we figured out